mysys/lf_alloc-pin.c

```diff
@@ -256,8 +256,8 @@ static int ptr_cmp(void **a, void **b)
 #define add_to_purgatory(PINS, ADDR)                                    \
   do                                                                    \
   {                                                                     \
-    *(void **)((char *)(ADDR)+(PINS)->pinbox->free_ptr_offset)=         \
-      (PINS)->purgatory;                                                \
+    my_atomic_storeptr_explicit((void **)((char *)(ADDR)+(PINS)->pinbox->free_ptr_offset), \
+      (PINS)->purgatory, MY_MEMORY_ORDER_RELEASE);                      \
```

Comment:
@grooverdan as soon as you accepted my suggestion, this one has to be `RELAXED` as well. Proper order in delete procedure is going to be enforced via `CAS(link, SEQ_CST)`.

Reply:
As far as I can see, `my_atomic_casptr` is _not_ a full barrier like `atomic_thread_fence(SEQ_CST)`. In Arm, for example, that would be compiled to something like this:

```
L_CAS: LDAXR X0, [link]   // my_atomic_casptr
            // if X0 different than expected branch to L_END
            STLXR X1, X2, [link]
            CBZ X1, L_CAS
L_END: // end of CAS
            // code continues
           // ...
L_KEY_UPDATE:  // my_atomic_storeptr_explicit((void **)((char *)(ADDR)+(PINS)->pinbox->free_ptr_offset)
            STR zero, [key]
```

According to the Arm memory model the STR can be reordered with STLXR, even though there exists a control dependency.

So I think you have to have either `STLR zero, [key]`, that is, the store to key is `MY_MEMORY_ORDER_RELEASE`, or you can add a full fence after the CAS to `link`, to ensure that nothing afterwards is moved upwards.

------------------------

That is why Linux always adds such a fence in RMW operations in their atomics (LKMM). Here is how Linux defines an cmpxchg for Armv8 without LSE:
```
	"1:	ld" #acq "xr" #sfx "\t%" #w "[oldval], %[v]\n"		\
	"	eor	%" #w "[tmp], %" #w "[oldval], %" #w "[old]\n"	\
	"	cbnz	%" #w "[tmp], 2f\n"				\
	"	st" #rel "xr" #sfx "\t%w[tmp], %" #w "[new], %[v]\n"	\
	"	cbnz	%w[tmp], 1b\n"					\
	"	" #mb "\n"						\
	"2:")								\
```
The `#mb` will be `dmb ish`, which is a full fence.

(code from here: https://elixir.bootlin.com/linux/latest/source/arch/arm64/include/asm/atomic_ll_sc.h#L279)

Reply:
@mbeck21 MariaDB works in terms of abstract memory models defined by C++: https://en.cppreference.com/w/cpp/atomic/memory_order.

Linux kernel doesn't follow it, which makes it inapplicable for MariaDB codebase. Terms like "full memory barrier", "read memory barrier", "write memory barriers" are inapplicable for C++ models as well. C++ valid models are "relaxed", "release-consume", "release-acquire" and "sequentially consistent".

According to this description `CAS(link, SEQ_CST)` is supposed to issue both `RELEASE` and `ACQUIRE`. And `ACQUIRE` won't let `key` get reordered before `CAS()` itself. Right?

Reply:
The relaxed store cannot be moved before the LDAXR of the CAS but before the STLXR part of the CAS, which is an issue that needs to be prevented. And making the store release does this.

Reply:
But then `CAS()` cannot be considered atomic anymore. Even if some architectures may split it in multiple steps, `ACQUIRE` must guarantee that all of these steps are completed before subsequent memory accesses.

Can it happen so that compiler emits something extra to enforce this requirement? Or probably some extra meaning is attached to existing instructions?

It'd be really unfortunate if some architectures were issuing barriers at different steps of their atomic operations.

Reply:
Regarding the atomicity of CAS or RMW.

Unfortunately, your assumption is not completely correct. The CAS is still atomic with respect to the variable it modifies (ie, `link`). But that is not the same as enforcing the order relation to operations on different variables.

The fact that CAS (and any RMW) is made out of two events (a reading and a writing) is not exclusive to Arm. The C11 memory model also allows the same kind of behavior, although the explanation wouldn't be based on "reordering".

To demonstrate that, have a look at a simple example to be run with [CppMem](http://svr-pes20-cppmem.cl.cam.ac.uk/cppmem), a web-based tool that generates executions possible within the c/c++ memory model (maybe you already know the tool).

```c
int main() {
    atomic_int key = 0;
    atomic_int link = 0; 
    int old = 0;
    {{{ { /* thread 1 */
          cas_strong_explicit(link, old, 1, memory_order_seq_cst, memory_order_seq_cst);
          key.store(1, memory_order_relaxed);
        }
    ||| { /* thread 2 */
          r1 = key.load(memory_order_seq_cst).readsvalue(1);
          r2 = link.load(memory_order_seq_cst).readsvalue(0);
        }
    }}};
    return 0; }
  ```

Now if you run this, you'll find it allows a _consistent_ execution of both threads such that the `key` update (`readsvalue(1)`) is seen but the update of `link` is not (`readsvalue(0)`), which is the same issue that is possible in the Arm memory model.

Note that the loads in thread 2 are issuing *memory_order_seq_cst* to separate the different concerns and avoid issues on this thread.

To fix this and make the bad execution impossible, we have to either make the store to `key` *memory_order_release*, or to add an `atomic_thread_fence(memory_order_release)` between the CAS to `link` and the store to `key`.

Reply:
For informational purposes.
A [similar discussion](https://gcc.gnu.org/legacy-ml/gcc/2014-02/msg00058.html) occurred in the GCC mailing list a few years back. CppMem was also used to confirm the behavior .

Reply:
@mbeck21 can the `key` update be seen before `link` update? Absolutely, because your example contains sequentially consistent model violation. It requires all involved atomic operations to be tagged with `seq_cst`, and in your example `key.store()` violates this requirement.

We discuss slightly different situation. Thread 2 should do:
```C++
r1 = key.load(memory_order_relaxed);
pins[0].store(1, memory_order_seq_cst); // release semantics matter here
if (link.load(memory_order_seq_cst) == 0)
  assert(r1 != 1); /* must never fire */
```

I'm more worried about cascading "release" semantics from `pins[0]` to `link`, but IIRC sequentially consistent allows this.

Now back to our original discussion re release-acquire model. You say that regular store can be reordered before store issued by multi-step `CAS()`. But then this commonly used trivial spin-lock implementation wouldn't be valid:
```C++
old= UNLOCKED;
while (!lock.compare_exchange_strong(old, LOCKED, std::memory_order_acquire)) { old= UNLOCKED; }
a= 1;
lock.store(UNLOCKED, std::memory_order_release);
```

Because you say that store to `a` can be reordered before store to `lock`. And to make sure it works properly one has to add extra `release`:
```C++
old= UNLOCKED;
while (!lock.compare_exchange_strong(old, LOCKED, std::memory_order_acquire)) { old= UNLOCKED; }
std::atomic_thread_fence(std::memory_order_release);
a= 1;
lock.store(UNLOCKED, std::memory_order_release);
```

Did I get your assertion correctly?

Reply:
It depends on how your example spin-lock is used. Lets assume that the code you pasted is running on thread 1.

### Case 1)
- Thread 2 is using the same `while(!cas)` pattern to lock and access `a`

In this case, some thread (e.g. Thread 1)  issues acquire on `lock`, succeeds in locking.
For Thread 1 it is possible that the relaxed store of `a` is performed before the store of `lock` (in the spin-loop).
However, as all other threads are accessing `a` via the same locking pattern, none of the other threads that try to acquire the lock will succeed as long as Thread 1 holds the lock. As long as the CAS fails, there is no valid path that leads to an access of `a`.

Using the spin-lock this way does not yield the bug mentioned above.

Observe, this is not the pattern happening in lf-hash with `key`. Threads are not locking access to `key` symmetrically this way. It does not apply to our discussion.

### Case 2)
- accesses to `a` are atomic relaxed in other threads, without the locking done by the CAS spin-loop

Other threads can observe the store of `a` before the store of `lock`. See this example program for cppmem:

```c
int main() {
    atomic_int lock = 0;
    atomic_int a = 0;
    int old = 0;
    {{{ { /* thread 1 */
      cas_strong_explicit(lock, old, 1, memory_order_acquire, memory_order_acquire);
      a.store(1, memory_order_relaxed);
      lock.store(2, memory_order_release);        
    }
    ||| { /* thread 2 */
          r2 = a.load(memory_order_relaxed).readsvalue(1);
          atomic_thread_fence(memory_order_seq_cst);
          r1 = lock.load(memory_order_relaxed).readsvalue(0);
        }
    }}};
    return 0; }
```

Thread 1 stores `2` for unlock to be sure that we read the initial unlocked state of `lock` in thread 2.
This yields a valid execution in the c++ memory model, meaning that indeed the store of `a` (reading `1`) is visible before the acquire of the lock (reading `0`).

### Case 3)
- `a` is non-atomic and read without locking 

There is a data race on `a` and the behavior is undefined. We can just ignore this case.

---

The write of `key` in `add_to_purgatory`, preceded by the CAS in `l_delete` to take the link off the list, while the other thread reads `key` in `l_find` is similar to case 2. In that case the store of `key` needs to be protected by an implicit barrier (atomic release store) in `add_to_purgatory`. As you have mentioned in a [previous](#discussion_r756825494) comment. 

I hope this answers your question.

Reply:
I meant case 1). If store of `a` was performed before store of `lock` in thread 1, why can't it get similar reordering in thread 2? And then in both threads store of `a` is going to happen while mutex is not fully locked.

So `CAS()` on ARM is "load-compare-store", I thought your original claim was that it may get reordered before "compare-store". Now I feel like you say it can only get reordered before "store". Then it'd probably work as expected.

Although there's also `compare_exchange()`, which can be used for mutex implementation instead of `CAS()`. And it doesn't have "compare" part.

Anyway now I got convinced that release-acquire has to be applied for `key` in another thread. Arguing with you was a pleasure! Hope time spent for this discussion is not a waste on your side either.

Reply:
Have you both reached a consensus yet? Do you want to create a new PR or a PR to my branch for the difference to prevent a(nother) mistake on my part. I'll ensure the 10.5 jump to c++ includes the type (to make windows happy and any other minor difference).

Reply:
@grooverdan, yes, I think so. Just issue `MY_MEMORY_ORDER_ACQUIRE` along with the `key` load in `l_find` and we can call it a day.

Reply:
@svoj Basically, the compare of the CAS determines if the out-of-order execution will be visible or not.
Thank you for being thorough on understanding the issue at hand. I also enjoyed our discussion. Time was certainly not wasted. :)